src/Agent.Worker/TestResults/Legacy/TestRunPublisher.cs

```diff
@@ -167,8 +207,117 @@ public TestRunData ReadResultsFromFile(TestRunContext runContext, string filePat
         }
+
+        public bool IsTestRunFailed(string projectName,
+           int testRunId,
+           CancellationToken cancellationToken)
+        {
+            Trace.Entering();
+            _publisherInputValidator.CheckTestRunId(testRunId);
+
+            _executionContext.Output(string.Format(CultureInfo.CurrentCulture, $"TestRunPublisher.IsTestRunFailed: Getting test run summary with run id: {testRunId} using stats api"));
+            var stats = _testResultsServer.GetTestRunStatisticsAsync(projectName, testRunId, cancellationToken: cancellationToken).Result;
+            return IsTestRunFailed(stats, cancellationToken);
+        }
+
+        public bool IsTestRunFailed(TestRunStatistic testRunStatistics,
+            CancellationToken cancellationToken)
+        {
+            Trace.Entering();
+
+            _executionContext.Output(string.Format(CultureInfo.CurrentCulture, "TestRunPublisher.IsTestRunFailed: checking if test run is failed using run summary input"));
+            if (testRunStatistics != null && testRunStatistics.RunStatistics?.Count > 0)
+            {
+                foreach (var stat in testRunStatistics.RunStatistics)
+                {
+                    if (stat.Outcome == TestOutcome.Failed.ToString() && stat.Count > 0 && stat.ResultMetadata == ResultMetadata.Flaky)
+                    {
+                        _executionContext.Output(string.Format(CultureInfo.CurrentCulture, "Number of flaky failed tests is: {0}", stat.Count));
+                    }
+
+                    if (stat.Outcome == TestOutcome.Failed.ToString() && stat.Count > 0 && stat.ResultMetadata != ResultMetadata.Flaky)
+                    {
+                        _executionContext.Output(string.Format(CultureInfo.CurrentCulture, "Number of failed tests which are non-flaky is: {0}", stat.Count));
+                        return true;
+                    }
+                }
+            }
+
+            Trace.Leaving();
+
+            return false;
+        }
+
+        public async Task<TestRunStatistic> GetTestRunSummaryAsync(string projectName,
+           int testRunId,
+           bool allowRetry,
+           CancellationToken cancellationToken)
+        {
+            Trace.Entering();
+
+            TestRunStatistic testRunStatistic = null;
+
+            _executionContext.Debug(string.Format(CultureInfo.CurrentCulture, $"TestRunPublisher.GetTestRunSummaryAsync: Getting test run summary with run id: {testRunId} using new summary api. Allow Retry: {allowRetry}"));
+
+            if (allowRetry)
+            {
+                var retryHelper = new RetryHelper(_executionContext, MaxRetries);
+
+                testRunStatistic = await retryHelper.Retry(() => GetTestRunSummaryByOutcome(projectName, testRunId, cancellationToken),
+                                                           (retryCounter) => GetTimeIntervalDelay(retryCounter),
```

Comment:
why new methods are added here? why can't they be in the new helper .cs file you have added TestRunDataPublisherhelper?

Reply:
Hope these methods are called inside the try catch block of the caller

Reply:
@navin22 From Design point of view it makes more sense to put this one there to me due to single responsibility principle all the interaction with server is happening in this class and sheer purpose of creating TestRunDataPublisherhelper class was to achieve code reusability for whatever common thing is happening between legacy and new code paths.

Reply:
These method looks common to both lib part and legacy part, but the code is in legacy ? why are these methods are only in legacy ?

Reply:
This is only for legacy path for lib path it uses GetTestRunSummaryAsync From library.